File: Data Structures & Algorithms/depth-of-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:

        def depth(node, length):
            if not node:
                return length
            length +=1
            return max(depth(node.left,length), depth(node.right,length))
        return depth(root, 0)
        

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private static int max = 0;
    private int calc(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = calc(node.left);
        int right = calc(node.right);
        max = Math.max(max, left + right);
        return 1 + Math.max(left, right);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        max = 0;
        int temp = calc(root);
        return max;
    }
}
